### PR TITLE
Zeus Mender integration for imx8mnevk

### PR DESCRIPTION
**Pull request by drewmoseley**
Zeus Mender integration for imx8mnevk

This currently requires using [my fork of imx-manifest.](https://github.com/drewmoseley/imx-manifest/tree/wip-zeus)

### CODE REVIEW

File: meta-mender-imx/README.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
+# meta-mender-imx
+
+Mender integration for imx based boards
+
+The supported and tested boards are:
+
+ - [i.MX8M Nano EVK](TBD)
```
Comment (mirzak) [inline meta-mender-imx/README.md:7]: It is this one right? https://hub.mender.io/t/nxp-i-mx-8m-mini-evaluation-kit/659

File: meta-mender-imx/README.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
+# meta-mender-imx
+
+Mender integration for imx based boards
```
Comment (mirzak) [inline meta-mender-imx/README.md:3]: Suggestion:

```
Mender integration for NXP i.MX BSP based boards
```

File: meta-mender-imx/README.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+URI: https://github.com/Freescale/meta-freescale-distro
+branch: zeus
+revision: HEAD
+```
+
+```
+URI: https://github.com/Freescale/meta-freescale
+branch: zeus
+revision: HEAD
+```
```
Comment (mirzak) [inline meta-mender-imx/README.md:33]: As this is based on the NXP BSP, there is a lot more to it here.

I would just say,

```
This layer depends on NXP BSP `imx-5.4.3-2.0.0`
```

File: meta-mender-imx/README.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+
+```
+mkdir mender-imx && cd mender-imx
+repo init -u https://github.com/mendersoftware/meta-mender-community \
+          -m meta-mender-imx/scripts/imx-5.4.3-2.0.0_demo_mender.xml \
+          -b zeus
+repo sync
+source setup-environment imx
+MACHINE=imx8mnevk bitbake core-image-base
```
Comment (mirzak) [inline meta-mender-imx/README.md:48]: I think there is a missing step here, I did something like this for the board I am working on,

```
mkdir mender-tn-imx-bsp && cd mender-tn-imx-bsp
repo init -u https://github.com/TechNexion/tn-imx-yocto-manifest.git -b zeus_5.4.y-next -m imx-5.4.24-2.1.0.xml
wget --directory-prefix .repo/manifests https://raw.githubusercontent.com/mendersoftware/meta-mender-community/zeus/meta-mender-tn-imx-bsp/scripts/imx-5.4.24-2.1.0_demo_mender.xml
repo init -m imx-5.4.24-2.1.0_demo_mender.xml
repo sync -j$(nproc)
DISTRO=fsl-imx-xwayland MACHINE=pico-imx8mm source tn-setup-mender.sh -b build-xwayland-imx8mm
bitbake core-image-base
```

File: meta-mender-imx/conf/layer.conf [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+            ${LAYERDIR}/recipes-*/*/*.bbappend"
+
+BBFILE_COLLECTIONS += "mender-imx"
+BBFILE_PATTERN_mender-imx := "^${LAYERDIR}"
+LAYERSERIES_COMPAT_mender-imx = "zeus"
+
+# This needs to be higher priority than meta-imx
+# We have to override the do_compile() function it provides for our newer
+# u-boot version.
+BBFILE_PRIORITY_mender-imx = "11"
```
Comment (mirzak) [inline meta-mender-imx/conf/layer.conf:15]: Nitpick, add `LAYERDEPENDS_`

File: meta-mender-imx/recipes-bsp/u-boot-imx/files/imx8mnevk/0001-Switch-to-CONFIG_DISTRO_DEFAULTS-for-bootcmd.patch [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
++	"run distro_bootcmd;"
+ 
+ /* Link Definitions */
+ #define CONFIG_LOADADDR			0x40480000
+@@ -222,7 +173,7 @@
+ #define CONFIG_ENV_SPI_MODE		CONFIG_SF_DEFAULT_MODE
+ #define CONFIG_ENV_SPI_MAX_HZ		CONFIG_SF_DEFAULT_SPEED
+ 
+-#define CONFIG_ENV_SIZE			0x1000
++#define CONFIG_ENV_SIZE			SZ_8K
```
Comment (mirzak) [inline meta-mender-imx/recipes-bsp/u-boot-imx/files/imx8mnevk/0001-Switch-to-CONFIG_DISTRO_DEFAULTS-for-bootcmd.patch:131]: Was it required to increase the env to fit distro_boot_cmd?

File: meta-mender-imx/scripts/imx-setup-mender.sh [hunk 7]
```diff
@@ -0,0 +1,31 @@
+# Mender Integration Setup Script
+#
+# Copyright 2019 NXP
+#
```
Comment (mirzak) [inline meta-mender-imx/scripts/imx-setup-mender.sh:4]: Nitpick, would maybe update this.

File: meta-mender-imx/templates/local.conf.append [hunk 8]
```diff
@@ -0,0 +1,18 @@
+
+# Appended fragment from meta-mender-community/meta-mender-imx/templates
+IMAGE_CLASSES += "mender-setup-imx"
+
+# General Settings
+MENDER_BOOT_PART_SIZE_MB_imx8mnevk = "256"
```
Comment (mirzak) [inline meta-mender-imx/templates/local.conf.append:6]: Why so large? I would assume that we actually can disable the boot part completely on this board?

File: meta-mender-imx/templates/local.conf.append [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+# Boot from SD
+#MENDER_STORAGE_DEVICE_imx8mnevk = "/dev/mmcblk1"
+
+# Boot from eMMC
+MENDER_STORAGE_DEVICE_imx8mnevk = "/dev/mmcblk2"
+
+IMAGE_FSTYPES_remove = "tar.bz2 ext4 sdcard.bz2 wic.bmap wic.bz2 uefiimg.bz2"
+IMAGE_BOOT_FILES_remove_imx8mnevk = " ${KERNEL_IMAGETYPE} ${@make_dtb_boot_files(d)} "
+MENDER_IMAGE_BOOTLOADER_FILE_imx8mnevk = "imx-boot"
+
```
Comment (mirzak) [inline meta-mender-imx/templates/local.conf.append:18]: These could probably be in `mender-setup-imx.bbclass` but not critical.

File: meta-mender-imx/README.md (hunk 1 above)
Comment (drewmoseley) [reply meta-mender-imx/README.md:7]: No.  That's the mini.  I have not created the hub page for it yet.

File: meta-mender-imx/README.md (hunk 4 above)
Comment (drewmoseley) [reply meta-mender-imx/README.md:48]: I am assuming that we will get the mender specifics into the imx-manifest repo.  If that doesn't happen then we will need the local_manifest stuff.

File: meta-mender-imx/recipes-bsp/u-boot-imx/files/imx8mnevk/0001-Switch-to-CONFIG_DISTRO_DEFAULTS-for-bootcmd.patch (hunk 6 above)
Comment (drewmoseley) [reply meta-mender-imx/recipes-bsp/u-boot-imx/files/imx8mnevk/0001-Switch-to-CONFIG_DISTRO_DEFAULTS-for-bootcmd.patch:131]: Yes

File: meta-mender-imx/templates/local.conf.append (hunk 8 above)
Comment (drewmoseley) [reply meta-mender-imx/templates/local.conf.append:6]: Not for GRUB-EFI integration.

Comment (drewmoseley) [reply meta-mender-imx/templates/local.conf.append:6]: Also, the optee and M4 bins are by default in /boot.  That may be changeable but I've not investigated yet.

Comment (mirzak) [reply meta-mender-imx/templates/local.conf.append:6]: > Not for GRUB-EFI integration.

Ah, true.

File: meta-mender-imx/README.md (hunk 2 above)
Comment (drewmoseley) [reply meta-mender-imx/README.md]: Fixed

File: meta-mender-imx/README.md (hunk 3 above)
Comment (drewmoseley) [reply meta-mender-imx/README.md]: Fixed

File: meta-mender-imx/scripts/imx-setup-mender.sh (hunk 7 above)
Comment (drewmoseley) [reply meta-mender-imx/scripts/imx-setup-mender.sh]: Fixed

File: meta-mender-imx/templates/local.conf.append (hunk 9 above)
Comment (drewmoseley) [reply meta-mender-imx/templates/local.conf.append]: Done

**Comment by drewmoseley** [comment]
@mirzak all comments addressed.

**Comment by drewmoseley** [comment]
I added a commit that I _think_ will support both imx8mq and imx8mm boards but I don't have hardware to test.
@mirzak can you do a boot time smoke test on either of those platforms?

File: meta-mender-imx/scripts/imx-5.4.3-2.0.0_demo_mender.xml [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+<manifest>
+
+  <include name="imx-5.4.3-2.0.0.xml"/>
+
+  <remote fetch="https://github.com/mendersoftware" name="mender"/>
+  <remote fetch="https://github.com/drewmoseley" name="drewmoseley"/>
+
+  <project name="meta-mender" remote="mender" revision="zeus" path="sources/meta-mender"/>
+  <project name="meta-mender-community" remote="drewmoseley" revision="wip-zeus-imx" path="sources/meta-mender-community">
+     <linkfile src="meta-mender-imx/scripts/imx-setup-mender.sh" dest="imx-setup-mender.sh"/>
```
Comment (mirzak) [inline meta-mender-imx/scripts/imx-5.4.3-2.0.0_demo_mender.xml:11]: Noticed that we called this one `fsl-setup-mender` on `sumo` branch. Might as well stick to that?

File: meta-mender-imx/README.md [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+that have Mender integrated.
+
+
+```
+mkdir mender-imx && cd mender-imx
+repo init -u https://github.com/mendersoftware/meta-mender-community \
+          -m meta-mender-imx/scripts/imx-5.4.3-2.0.0_demo_mender.xml \
+          -b zeus
+repo sync
+source setup-environment imx
```
Comment (mirzak) [inline meta-mender-imx/README.md:38]: This should be something like:

```
DISTRO=fsl-imx-xwayland MACHINE=imx8mqevk . imx-setup-mender.sh -b build
```

**Comment by mirzak** [comment]
> @mirzak can you do a boot time smoke test on either of those platforms?

I could test on the `imx8mqevk`. 

Though I see one problem. As we supported `imx8mqevk` and `imx8mmevk` on the `sumo` branch using a regular U-Boot integration, and this updated integration now uses GRUB, we will not support an upgrade path on these boards from `sumo` to `zeus`.

I do not know how many users we have of these integrations, but would still like to avoid breaking the update path if possible.

File: meta-mender-imx/scripts/imx-5.4.3-2.0.0_demo_mender.xml (hunk 10 above)
Comment (drewmoseley) [reply meta-mender-imx/scripts/imx-5.4.3-2.0.0_demo_mender.xml:11]: This is based on the rename that happened in meta-imx.  I am trying to be consistent with that.

**Comment by drewmoseley** [comment]
It seems likely that big of a jump will have other issues.

File: meta-mender-imx/README.md (hunk 11 above)
Comment (drewmoseley) [reply meta-mender-imx/README.md]: Fixed.

**Comment by mirzak** [comment]
> It seems likely that big of a jump will have other issues.

I can not really see any issues with OTA capabilities in mind, and the changes here would intentionally break that which I would like to avoid if possible.

**Comment by mirzak** [comment]
To avoid blocking this PR, maybe we can just merge the `imx8mnevk` bits while we make a final decision on the other types?

**Comment by drewmoseley** [comment]
Works for me.

**Pull request merged**